### PR TITLE
Nearest neighbor interpolator

### PR DESCRIPTION
**Pull request by nbren12**

### CODE REVIEW

**Comment by nbren12** [comment]
Nearest neighbor
image

**Comment by nbren12** [comment]
Inverse distance
image

File: earth2grid/_regrid.py [hunk 1]
```diff
@@ -48,22 +50,216 @@ def forward(self, x):
...
+        """
+        *shape, x = z.shape
+        zrs = z.view(-1, x).T
+        # using embedding bag is 2x faster on cpu and 4x on gpu.
+        output = torch.nn.functional.embedding_bag(self.index, zrs, mode='sum')
+        output = output.T.view(*shape, -1)
+        return output
+
+
+class S2LinearBarycentricInterpolator(torch.nn.Module):
```
Comment (akshaysubr) [inline earth2grid/_regrid.py:203]: Would be good to add a test that saves a checkpoint of an initialized instance of this and loads that back in. Would be very useful to avoid repeated calls to `spatial.KDTree` especially in ETL jobs where a regridding script might be invoked multiple times on different inputs.

File: earth2grid/_regrid.py [hunk 2]
```diff
@@ -48,22 +50,216 @@ def forward(self, x):
...
+        dest_lat = torch.deg2rad(dest_lat.cpu())
+
+        vec = torch.stack(ang2vec(src_lon, src_lat), -1)
+
+        # havesign distance and euclidean are monotone for points on S2 so can use 3d lookups.
+        self.tree = spatial.KDTree(vec)
+        vec = torch.stack(ang2vec(dest_lon.cpu(), dest_lat.cpu()), -1)
+        _, neighbors = self.tree.query(vec, k=3)
+        d = haversine_distance(dest_lon[:, None], dest_lat[:, None], src_lon[neighbors], src_lat[neighbors])
+        lam = 1 / d
```
Comment (akshaysubr) [inline earth2grid/_regrid.py:240]: Should probably regularize this with an epsilon in the denominator

File: tests/test_regrid.py [hunk 3]
```diff
@@ -195,3 +195,35 @@ def test_out_of_bounds():
...
+    expected = torch.cos(torch.deg2rad(lond))
+    mae = torch.mean(torch.abs(out - expected))
+    assert mae.item() < 0.02
+
+
+def test_BaryCentric():
+    n = 10000
+    torch.manual_seed(0)
+    lon = torch.rand(n) * 360
+    lat = torch.rand(n) * 180 - 90
```
Comment (akshaysubr) [inline tests/test_regrid.py:220]:
```suggestion
    lat = torch.asin(2*torch.rand(n)-1) * 180. / np.pi
```

**Review by akshaysubr — APPROVED** [review]
Looks good! Had a couple of comments about testing saving the interpolator and one minor comment on sampling for the tests.

File: earth2grid/_regrid.py (hunk 1 above)
Comment (nbren12) [reply earth2grid/_regrid.py:203]: I'll make an issue for this. I think it may require some thought and potentially backwards breaking changes to apply this consistently across the suite of re-gridders.

I'm not sure the most elegant way to allow ser/deser of "buffer-only" modules like this. Do you know of an example for this e.g. with normalization.

Comment (akshaysubr) [reply earth2grid/_regrid.py:203]: I think we don't necessarily need a proper serialization/deserialization pipeline here right? It might be sufficient to create a classmethod that takes in a state dict and returns an initialized instance from that?

Comment (nbren12) [reply earth2grid/_regrid.py]: I refactored this so that the `S2NearestNeighborInterpolator` now returns a serializeable `Regridder` object. Let me know what you think.

File: tests/test_regrid.py (hunk 3 above)
Comment (nbren12) [reply tests/test_regrid.py]: what is the intent here?

Comment (akshaysubr) [reply tests/test_regrid.py]: This isn't a super important change, but thought it would be good to remove the polar bias when sampling on the sphere in these tests.

File: earth2grid/_regrid.py [hunk 4]
```diff
@@ -12,12 +12,48 @@
...
+        weight = self.weight.view(-1, p)
 
-from earth2grid import base, healpix
-from earth2grid.latlon import LatLonGrid
+        # using embedding bag is 2x faster on cpu and 4x on gpu.
+        output = torch.nn.functional.embedding_bag(index, zrs, per_sample_weights=weight, mode='sum')
+        output = output.T.view(*shape, -1)
+        return output.reshape(list(shape) + output_shape)
+
+    @staticmethod
```
Comment (akshaysubr) [inline earth2grid/_regrid.py:51]: Shouldn't this be a `classmethod`?

File: earth2grid/_regrid.py [hunk 5]
```diff
@@ -48,22 +84,147 @@ def forward(self, x):
...
+        output = torch.nn.functional.embedding_bag(self.index, zrs, per_sample_weights=self.weights, mode='sum')
+        interpolated = torch.full(
+            [self.mask.numel(), zrs.shape[1]], fill_value=self.fill_value, dtype=z.dtype, device=z.device
+        )
+        interpolated.masked_scatter_(self.mask.unsqueeze(-1), output)
+        interpolated = interpolated.T.view(*shape, self.mask.numel())
+        return interpolated
+
+
+def S2NearestNeighborInterpolator(
```
Comment (akshaysubr) [inline earth2grid/_regrid.py:185]: Name might be a bit misleading with the unification? Might be better to call it BarycentricInterpolator and can maybe have NearestNeighbor be a specialization of this with `k = 1` or just note this `k = 1` behavior in the docs?

**Review by akshaysubr — APPROVED** [review]
Changes look good and the unification of the two KDTree implementations is nice too.

Not for this PR, but might be nice to have the `Regridder` object be TorchScript compatible so you could serialize a regridder object that way.

File: tests/test_regrid.py [hunk 6]
```diff
@@ -195,3 +195,23 @@ def test_out_of_bounds():
...
+
+
+@pytest.mark.parametrize("k", [1, 2, 3])
+def test_NearestNeighborInterpolator(k):
+    n = 10000
+    torch.manual_seed(0)
+    lon = torch.rand(n) * 360
+    lat = torch.rand(n) * 180 - 90
+
+    lond = torch.rand(n) * 360
```
Comment (akshaysubr) [inline tests/test_regrid.py:207]: Would also be good to test with a different number of source and destination query points

File: earth2grid/_regrid.py [hunk 7]
```diff
@@ -48,22 +84,147 @@ def forward(self, x):
...
+    src_lon: torch.Tensor,
+    src_lat: torch.Tensor,
+    dest_lon: torch.Tensor,
```
Comment (akshaysubr) [inline earth2grid/_regrid.py:188]: Would be good to add a check that this is a 1D array or reshape a 2D array to 1D if needed.

File: tests/test_regrid.py [hunk 8]
```diff
@@ -195,3 +195,23 @@ def test_out_of_bounds():
...
+    latd = torch.rand(n) * 180 - 90
+
+    interpolate = earth2grid.S2NearestNeighborInterpolator(lon, lat, lond, latd, k=k)
+    out = interpolate(torch.cos(torch.deg2rad(lon)))
```
Comment (akshaysubr) [inline tests/test_regrid.py:211]: Might also be good to check for batched interpolation since the Regridder seems to support that?

File: earth2grid/_regrid.py (hunk 4 above)
Comment (nbren12) [reply earth2grid/_regrid.py:51]: I don't intend to inherit this object so used staticmethod and explicitly refer to the class name instead:
```
        n, p = d["index"].shape
        regridder = Regridder((n,), p)
        regridder.load_state_dict(d)
```

File: earth2grid/_regrid.py (hunk 5 above)
Comment (nbren12) [reply earth2grid/_regrid.py:185]: I realized it's not actually barycentric renaming to `KNNInterpolatorS2`.

Comment (nbren12) [reply earth2grid/_regrid.py:185]: new name is actually `KNNS2Interpolator`

File: tests/test_regrid.py (hunk 6 above)
Comment (nbren12) [reply tests/test_regrid.py:207]: done

File: earth2grid/_regrid.py (hunk 7 above)
Comment (nbren12) [reply earth2grid/_regrid.py:188]: added an error.

File: tests/test_regrid.py (hunk 8 above)
Comment (nbren12) [reply tests/test_regrid.py:211]: done
